sdk/storage/AzureStorageBlob/Source/DataStructures/BlockOperations.swift

```diff
@@ -69,18 +61,33 @@ public class BlockOperation: ResumableTransfer {
                 endRange: Int(transfer.endRange),
                 options: downloader.options
             )
-            chunkDownloader.download { result, _ in
+            chunkDownloader.download { result, httpResponse in
+                if self.isCancelled || self.isPaused { return }
                 switch result {
                 case .success:
-                    self.delegate?.operation(self, didChangeState: .inProgress)
-                case .failure:
-                    self.transfer?.state = .failed
-                    // TODO: The failure needs to propagate to the entire operation...
-                    self.delegate?.operation(self, didChangeState: .failed)
+                    guard let responseHeaders = httpResponse?.headers else {
+                        assertionFailure("Response headers not found.")
+                        return
+                    }
+                    let blobProperties = BlobProperties(from: responseHeaders)
+                    let contentLength = blobProperties.contentLength ?? 0
+                    downloader.progress += contentLength
+                    transfer.state = .complete
+                    self.notifyDelegate(withTransfer: transfer)
+                case let .failure(error):
+                    var err = error
+                    if let pipelineError = error as? PipelineError {
+                        err = pipelineError.innerError
+                    }
+                    transfer.state = .failed
```

Comment:
Should `state` be a property that, when updated, contains the logic to update the parent as appropriate as well? I feel like it might be a good way to make sure that all the "state" properties always get updated correctly.

Reply:
It's certainly possible, but I'm not sure it would be a simple value add. There are states for which a change in the child shouldn't propagate up to the parent. It's reasonable to say that if any block failed, the whole blob transfer is a failure, but for example, when a block is completed, the blob download itself is very likely not.